<launch>

    <!-- 
        Author: Dr. -Ing. Ahmad Kamal Nasir
        Email: dringakn@gmail.com
        Desciption: Setup the navigation stack with the global and local planner.
                    For the global planner use A* and for the local planner use
                    DWA/TEB.
    -->

    <arg name="robot_odom_frame" default="odom"/>
    <arg name="robot_frame" default="base_link"/>
    <arg name="map_topic" default="/map"/>
    <arg name="map_frame" default="map"/>
    <arg name="navigation_global_map_resolution" default="0.1"/>
    <arg name="navigation_global_map_width" default="10"/>
    <arg name="navigation_global_map_length" default="10"/>
    <arg name="navigation_local_map_resolution" default="0.1"/>
    <arg name="navigation_local_map_width" default="10"/>
    <arg name="navigation_local_map_length" default="10"/>
    <arg name="footprint" default="[[0.25, -0.15], [0.25, 0.15], [-0.15, 0.15], [-0.15, -0.15]]"/>
    <arg name="footprint_padding" default="0"/>

    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">

        <!-- Costmap configurations -->
        <rosparam file="$(dirname)/../cfg/navigation_stack/common_costmap_params.yaml" command="load" ns="global_costmap"/>
        <rosparam file="$(dirname)/../cfg/navigation_stack/common_costmap_params.yaml" command="load" ns="local_costmap" />
        <rosparam file="$(dirname)/../cfg/navigation_stack/global_costmap_params.yaml" command="load"/>
        <rosparam file="$(dirname)/../cfg/navigation_stack/local_costmap_params.yaml" command="load"/>

        <!-- Global planner -->
        <param name="base_global_planner" value="global_planner/GlobalPlanner" />

        <!-- Local Planner ROS -->
        <!-- <rosparam file="$(dirname)/../cfg/navigation_stack/base_local_planner_params.yaml" command="load" />
        <param name="base_local_planner" value="local_planner/TrajectoryPlannerROS" /> -->

        <!-- Local Planner TEB -->
        <rosparam file="$(dirname)/../cfg/navigation_stack/teb_local_planner_params.yaml" command="load" />
        <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" />
        <param name="TebLocalPlannerROS/odom_topic" value="/$(arg robot_odom_frame)"/>
        <param name="TebLocalPlannerROS/map_frame" value="$(arg map_frame)"/>
        <param name="TebLocalPlannerROS/footprint_model/vertices" value="$(arg footprint)"/>

        <!-- Local Costmap Configurations -->
        <param name="local_costmap/robot_base_frame" value="$(arg robot_frame)"/>
        <param name="local_costmap/width" value="$(arg navigation_local_map_width)"/>
        <param name="local_costmap/height" value="$(arg navigation_local_map_length)"/>
        <param name="local_costmap/resolution" value="$(arg navigation_local_map_resolution)"/>
        <param name="local_costmap/global_frame" value="$(arg map_frame)"/>
        <param name="local_costmap/transform_tolerance" value="5"/>
        <param name="local_costmap/update_frequency" value="2"/>
        <param name="local_costmap/publish_frequency" value="2"/>
        <param name="local_costmap/rolling_window" value="true"/>
        <param name="local_costmap/origin_x" value="0"/>
        <param name="local_costmap/origin_y" value="0"/>
        <param name="local_costmap/always_send_full_costmap" value="true"/>
        <param name="local_costmap/static_layer/map_topic" value="$(arg map_topic)"/>
        <param name="local_costmap/footprint" value="$(arg footprint)"/>
        <!-- send_full_costmap: false -->

        <!-- Global Costmap Configurations -->
        <param name="global_costmap/robot_base_frame" value="$(arg robot_frame)"/>
        <param name="global_costmap/width" value="$(arg navigation_global_map_width)"/>
        <param name="global_costmap/height" value="$(arg navigation_global_map_length)"/>
        <param name="global_costmap/resolution" value="$(arg navigation_global_map_resolution)"/>
        <param name="global_costmap/global_frame" value="$(arg map_frame)"/>
        <param name="global_costmap/transform_tolerance" value="5"/>
        <param name="global_costmap/update_frequency" value="1"/>
        <param name="global_costmap/publish_frequency" value="1"/>
        <param name="global_costmap/rolling_window" value="true"/>
        <param name="global_costmap/origin_x" value="0"/>
        <param name="global_costmap/origin_y" value="0"/>
        <param name="global_costmap/always_send_full_costmap" value="true"/>
        <param name="global_costmap/static_layer/map_topic" value="$(arg map_topic)"/>
        <param name="global_costmap/footprint" value="$(arg footprint)"/>

        <!-- Planner Configurations -->
        <param name="controller_frequency" value="5"/>
        <param name="planner_frequency" value="0.5"/>
        <param name="controller_patience" value="10"/>
        <param name="planner_patience" value="10"/>
        <param name="max_planning_retries" value="-1"/>
        <param name="conservative_reset_dist" value="0.0"/>
        <param name="recovery_behavior_enabled" value="true"/>
        <param name="clearing_rotation_allowed" value="false"/>
        <param name="shutdown_costmaps" value="false"/>
        <param name="oscillation_timeout" value="30"/>
        <param name="oscillation_distance" value="0.5"/>
        <param name="restore_defaults" value="false"/>

    </node>

    <!-- <node pkg="rqt_reconfigure" type="rqt_reconfigure" name="rqt_reconfigure" required="true" /> -->

</launch>